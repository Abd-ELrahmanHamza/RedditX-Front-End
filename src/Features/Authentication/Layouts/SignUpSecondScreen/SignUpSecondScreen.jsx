/* eslint-disable react-hooks/exhaustive-deps */
import { useState, useEffect} from "react";

import { IoMdArrowBack } from "react-icons/io";
import { TfiReload } from "react-icons/tfi";

import ReCAPTCHA from "react-google-recaptcha";

import RandomUserName from "Features/Authentication/Utils/RandomUserName";

import FormInput from "Features/Authentication/Components/FormInput/FormInput";
import Button from "../../Components/Button/Button";
import PasswordStrength from "Features/Authentication/Components/PasswordStrength/PasswordStrength";
import GetPasswordStrength from "Features/Authentication/Utils/GetPasswordStrenght";

import {
  ButtonsContainer,
  ErrorParagraph,
  AuthParagraph,
  AuthHeader,
} from "../LogIn/LogIn.styled";

import {
  AuthContainer,
  BackSpan,
  Group,
  ReCAPTCHAContainer,
} from "./SignUpSecondScreen.styled";

const USER_NAME = /^[A-z][A-z0-9-_]{3,20}$/;
const PWD_REGEX = /^[A-z][A-z0-9-_]{8,20}$/;



/**
 * SignUpSecondScreen component that is used in Signup Component
 *  @param {boolean} initialFocus Prop to know if the user made at least one focus on the input field or not
 * @param {Function} setInitialFocus Function to set the state of initialFocus
 * @param {Function} setValidUserName Function to set the validity of the userName
 * @param {boolean} validUserName Prop to know if the userName is valid or not
 * @param {boolean} secondScreen Prop to know what screen should be shown (the choose email screen or choose userName and password screen)
 * @param {Function} setSecondScreen Function to set the state of secondScreen
 * @param {boolean} validPassword Prop to know if the password is valid or not
 * @param {Function} setValidPassword Function to set the validity of the password
 * @param {boolean} notRobot Prop to know if the user passed the captcha or not
 * @param {Function} setNotRobot Function to set the state of notRobot
 * @param {number} passwordStrength The strength of the password
 * @param {Function} setPasswordStrength Function to set the strength of the password
 * @param {Object} formFields Object contain the values of input fields
 * @param {Function} setFormFields Function to update the values of input fields
 * @returns {React.Component}  SignUpSecondScreen component that is used in Signup Component
 */

const SignUpSecondScreen = ({
  initialFocus,
  setInitialFocus,
  initialFocus2,
  setInitialFocus2,
  validUserName,
  setValidUserName,
  secondScreen,
  setSecondScreen,
  validPassword,
  setValidPassword,
  notRobot,
  setNotRobot,
  passwordStrength,
  setPasswordStrength,
  formFields,
  setFormFields
}) => {
 

  /**
   * state to know what error message should be shown
   */
  const [errMsg, setErrMsg] = useState("");
  const { userName, password } = formFields;

  /**
   * some configurations for the captcha
   */
  const grecaptchaObject = window.grecaptcha;
  window.recaptchaOptions = {
    useRecaptchaNet: true,
  };

 

  /**
   * useEffect for userName field to check if the userName that the user entered is valid or not
   */
  useEffect(() => {
    if (userName.length === 0) {
      setErrMsg("Username must be between 3 and 20 characters");
    }
    setValidUserName(USER_NAME.test(userName));
  }, [userName]);

  /**
   * useEffect for password field to check if the userName that the user entered is valid or not
   */
  useEffect(() => {
    if (password.length > 0) {
      setInitialFocus(false);
    }
    setValidPassword(PWD_REGEX.test(password));
  }, [password]);

  /**
   * Function to handle the submit of the form of signup
   * @param {*} event
   */
  const handleSubmit = async (event) => {
    event.preventDefault();
  };

  /**
   * Function to handle any change on the input field of the login form (check if the userName or the email or the password is valid or not)
   * @param {*} event
   */
  const handleChange = (event) => {
    const { name, value } = event.target;
    if (name === "password") {
      setPasswordStrength(GetPasswordStrength(password));
    }
    if (name === "userName") {
      setValidUserName(USER_NAME.test(userName));
    }
    setFormFields({ ...formFields, [name]: value });
  };

  /**
   * Function to change userName shown in the userName textField by calling RandomUserName (function in utilities folder)
   */
  const changeUserName = () => {
    setFormFields({ ...formFields, userName: RandomUserName() });
    setValidUserName(true);
  };

  /**
   * Function to handle the captcha and set the state of notRobot to true
   * @param {boolean} value the value of captcha
   */
  const recaptchaHandler = (value) => {
    setNotRobot(true);
  };

  return (
    <>
      {
        <AuthContainer secondScreen={!secondScreen}>
          <BackSpan
            onClick={() => {
              setSecondScreen(false);
            }}
          >
            <IoMdArrowBack size={20} />
          </BackSpan>

          <br></br>
          <br></br>
          <AuthHeader>Create your username and password</AuthHeader>
          <AuthParagraph>
            Reddit is anonymous, so your username is what you’ll go by here.
            Choose wisely—because once you get a name, you can’t change it.
          </AuthParagraph>

          <br></br>

          <form onSubmit={handleSubmit}>
            <Group>
              <FormInput
                valid={validUserName}
                initialFocus={initialFocus}
                label="Username"
                type="text"
                required
                onChange={handleChange}
                name="userName"
                value={userName}
                onFocus={() => {
                  setInitialFocus(false);
                }}
              />
              <span onClick={() => changeUserName()}>
                <TfiReload size={14} />
              </span>
            </Group>

            {/* Show error message if the userName is not valid and the user made a focus on the it's input field */}

            <ErrorParagraph valid={validUserName || initialFocus}>
              {errMsg}
            </ErrorParagraph>

            <Group>
              <FormInput
                valid={validPassword}
                initialFocus={initialFocus2}
                label="Password"
                type="password"
                required
                onChange={handleChange}
                name="password"
                value={password}
                onFocus={() => {
                  setInitialFocus2(false);
                }}
              />
              {!initialFocus2 && (
                <span onClick={() => {}}>
                  <PasswordStrength
                    strength={passwordStrength}
                  ></PasswordStrength>
                </span>
              )}
            </Group>

            {/* Show error message if the password is not valid and the user made a focus on the it's input field */}
            <ErrorParagraph valid={validPassword || initialFocus2}>
              password should contain 8 to 20 characters
            </ErrorParagraph>
            <ButtonsContainer>
              {/** Show an enabled button if the userName is valid and the password is valid and the user has passed the captcha*/}
              {validUserName && validPassword && notRobot && (
                <Button
                  valid={validUserName && validPassword && notRobot}
                  type="button"
                  onClick={() => RandomUserName()}
                >
                  Continue
                </Button>
              )}

              {/** Show an disabled button if the userName is not valid or the password is not valid or the user has not passed the captcha*/}
              {(!validUserName || !validPassword || !notRobot) && (
                <Button
                  disabled
                  valid={validUserName && validPassword && notRobot}
                  type="button"
                >
                  Continue
                </Button>
              )}
            </ButtonsContainer>
          </form>

          <ReCAPTCHAContainer validEmail={validUserName && validPassword}>
            <ReCAPTCHA
              sitekey="6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI"
              onChange={recaptchaHandler}
              grecaptcha={grecaptchaObject}
            />
          </ReCAPTCHAContainer>
        </AuthContainer>
      }
    </>
  );
};

export default SignUpSecondScreen;
