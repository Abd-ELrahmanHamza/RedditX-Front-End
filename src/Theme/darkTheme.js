/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
  },
  // The colors of the backgrounds in the website
  background: {
    primary: "#1a1a1b",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
  },
};

export default darkTheme;
