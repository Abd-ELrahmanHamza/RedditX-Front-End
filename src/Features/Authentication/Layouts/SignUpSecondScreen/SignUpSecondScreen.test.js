import { shallow } from "enzyme";

import SignUpSecondScreen from "./SignUpSecondScreen";
import { BrowserRouter, Routes } from "react-router-dom";

import { screen, render, fireEvent, waitFor } from "@testing-library/react";

import { ThemeProvider } from "styled-components";

import darkTheme from "Theme/darkTheme";

const validPassword = false;
const validUserName = false;
const userNameScreen = false;
const secondScreen = false;
const initialFocus = false;
const initialFocus2 = false;
const notRobot = false;
const passwordStrength = 5;
const setValidPassword = jest.fn();
const setModalShowLogIn = jest.fn();
const setFormFields = jest.fn();
const setUserNameScreen = jest.fn();
const setValidUserName = jest.fn();
const setSecondScreen = jest.fn();
const setInitialFocus = jest.fn();
const setInitialFocus2 = jest.fn();
const setErrMsg = jest.fn();
const setNotRobot = jest.fn();
const setPasswordStrength = jest.fn();
const setModalShowSignUp = jest.fn();
const setModalAfterSignUp = jest.fn();

const defaultFormFields = {
  userName: "",
  password: "",
  email: "",
};

describe("SignUpSecondScreen Layout", () => {
  it("this is a test for SignUpSecondScreen Layout", () => {
    expect(
      shallow(
        <BrowserRouter>
          <Routes>
            <SignUpSecondScreen
              setValidPassword={setValidPassword}
              setModalShowLogIn={setModalShowLogIn}
              setFormFields={setFormFields}
              setUserNameScreen={setUserNameScreen}
              setValidUserName={setValidUserName}
              setSecondScreen={setSecondScreen}
              setInitialFocus={setInitialFocus}
              setInitialFocus2={setInitialFocus2}
              setErrMsg={setErrMsg}
              setNotRobot={setNotRobot}
              setPasswordStrength={setPasswordStrength}
              setModalShowSignUp={setModalShowSignUp}
              setModalAfterSignUp={setModalAfterSignUp}
              formFields={defaultFormFields}
            />
          </Routes>
        </BrowserRouter>
      )
    ).toMatchSnapshot();
  });

  test("username-error shouldn't be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <SignUpSecondScreen
            setValidPassword={setValidPassword}
            setModalShowLogIn={setModalShowLogIn}
            setFormFields={setFormFields}
            setUserNameScreen={setUserNameScreen}
            setValidUserName={setValidUserName}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            setNotRobot={setNotRobot}
            setPasswordStrength={setPasswordStrength}
            setModalShowSignUp={setModalShowSignUp}
            setModalAfterSignUp={setModalAfterSignUp}
            formFields={defaultFormFields}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    expect(screen.queryByTestId("username-error")).not.toBeVisible();
  });


  test("username-error should be not found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <SignUpSecondScreen
            setValidPassword={setValidPassword}
            setModalShowLogIn={setModalShowLogIn}
            setFormFields={setFormFields}
            setUserNameScreen={setUserNameScreen}
            setValidUserName={setValidUserName}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            setNotRobot={setNotRobot}
            setPasswordStrength={setPasswordStrength}
            setModalShowSignUp={setModalShowSignUp}
            setModalAfterSignUp={setModalAfterSignUp}
            formFields={defaultFormFields}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    const userNameInput = screen.getByTestId("username");

    fireEvent.change(userNameInput, {
      target: { value: "Mgfdgs" },
    });

    const userNameError = screen.getByTestId("username-error");

    await waitFor(() => {
      expect(userNameError).not.toBeVisible();
    });
  });

  test("password-input should be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <SignUpSecondScreen
            validPassword={validPassword}
            validUserName={validUserName}
            userNameScreen={userNameScreen}
            secondScreen={secondScreen}
            initialFocus={initialFocus}
            initialFocus2={initialFocus2}
            notRobot={notRobot}
            passwordStrength={passwordStrength}
            setValidPassword={setValidPassword}
            setModalShowLogIn={setModalShowLogIn}
            setFormFields={setFormFields}
            setUserNameScreen={setUserNameScreen}
            setValidUserName={setValidUserName}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            setNotRobot={setNotRobot}
            setPasswordStrength={setPasswordStrength}
            setModalShowSignUp={setModalShowSignUp}
            setModalAfterSignUp={setModalAfterSignUp}
            formFields={defaultFormFields}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    const userNameInput = screen.getByTestId("password");
    expect(userNameInput).toBeInTheDocument();
  });
});
