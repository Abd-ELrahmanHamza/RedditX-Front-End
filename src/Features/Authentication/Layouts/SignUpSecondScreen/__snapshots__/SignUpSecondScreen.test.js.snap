// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpSecondScreen Layout this is a test for SignUpSecondScreen Layout 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "createHref": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Routes>
    <SignUpSecondScreen
      formFields={
        Object {
          "email": "",
          "password": "",
          "userName": "",
        }
      }
      setErrMsg={[MockFunction]}
      setFormFields={[MockFunction]}
      setInitialFocus={[MockFunction]}
      setInitialFocus2={[MockFunction]}
      setModalAfterSignUp={[MockFunction]}
      setModalShowLogIn={[MockFunction]}
      setModalShowSignUp={[MockFunction]}
      setNotRobot={[MockFunction]}
      setPasswordStrength={[MockFunction]}
      setSecondScreen={[MockFunction]}
      setUserNameScreen={[MockFunction]}
      setValidPassword={[MockFunction]}
      setValidUserName={[MockFunction]}
    />
  </Routes>
</Router>
`;
